File: app/src/main/java/me/srikavin/quiz/viewmodel/QuizDetailViewModel.java
package me.srikavin.quiz.viewmodel;

import android.app.Application;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import me.srikavin.quiz.model.Quiz;
import me.srikavin.quiz.repository.QuizRepository;

public class QuizDetailViewModel extends AndroidViewModel {

    private final QuizRepository quizRepository;
    private Map<String, MutableLiveData<Quiz>> quizzes = new HashMap<>();

    public QuizDetailViewModel(@NonNull Application application) {
        super(application);
        quizRepository = new QuizRepository(application);
    }

    public LiveData<Quiz> getQuizByID(String id) {
        if (!quizzes.containsKey(id)) {
            quizzes.put(id, new MutableLiveData<>());
            updateQuiz(id);
        }
        return quizzes.get(id);
    }

    public void updateQuiz(final String id) {
        quizRepository.getQuizByID(id, new QuizRepository.QuizResponseHandler() {
            @Override
            public void handle(@Nullable Quiz newQuiz) {
                quizzes.get(id).postValue(newQuiz);
            }
        });
    }
}

File: app/src/main/java/me/srikavin/quiz/viewmodel/QuizDetailViewModel.kt
package me.srikavin.quiz.viewmodel

import android.content.Context
import com.ww.roxie.BaseAction
import com.ww.roxie.BaseState
import com.ww.roxie.BaseViewModel
import com.ww.roxie.Reducer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.ofType
import io.reactivex.rxkotlin.plusAssign
import me.srikavin.quiz.model.Quiz
import me.srikavin.quiz.repository.QuizRepository

sealed class QuizDetailAction : BaseAction {
    data class LoadQuiz(val quizId: String) : QuizDetailAction()
}

sealed class QuizDetailChange {
    object Loading : QuizDetailChange()
    data class QuizLoad(val quiz: Quiz) : QuizDetailChange()
    data class Error(val error: Throwable?) : QuizDetailChange()
}

data class QuizDetailState(
        val quiz: Quiz? = null,
        val loading: Boolean = false,
        val error: Boolean = false
) : BaseState

class QuizDetailViewModel(initialState: QuizDetailState?, context: Context) : BaseViewModel<QuizDetailAction, QuizDetailState>() {
    private val quizRepository: QuizRepository = QuizRepository(context)
    override val initialState = initialState ?: QuizDetailState()

    private val reducer: Reducer<QuizDetailState, QuizDetailChange> = { state, change ->
        when (change) {
            is QuizDetailChange.Loading -> state.copy(
                    loading = true,
                    error = false
            )

            is QuizDetailChange.QuizLoad -> state.copy(
                    quiz = change.quiz,
                    loading = false,
                    error = false
            )

            is QuizDetailChange.Error -> state.copy(
                    error = true,
                    loading = false
            )
        }
    }

    init {
        bindActions()
    }

    private fun bindActions() {
        val quizDetailChanges = actions.ofType<QuizDetailAction.LoadQuiz>()
                .switchMap { action ->
                    quizRepository
                            .getQuizByID(action.quizId)
                            .observeOn(AndroidSchedulers.mainThread())
                            .toObservable()
                            .map<QuizDetailChange> { QuizDetailChange.QuizLoad(it) }
                            .onErrorReturn { QuizDetailChange.Error(it) }
                            .startWith(QuizDetailChange.Loading)
                }

        disposables += quizDetailChanges
                .scan(initialState, reducer)
                .distinctUntilChanged()
                .subscribe(state::postValue)
    }
}
